File: MovieTheater/MovieTheater.Domain/Entities/Movie.cs
namespace MovieTheater.Domain.Entities;

public class Movie
{
    public Guid Id { get; set; }
    public string Title { get; set; }
    public string Description { get; set; }
    public DateTime DateReleased { get; set; }
    public int DurationMinutes { get; set; }
    public decimal Budget { get; set; }
    public double IMDbRating { get; set; }
    public decimal Revenue { get; set; }
    public string Poster { get; set; }

    public Guid? MovieDirectorId { get; set; }
    public virtual MovieDirector MovieDirector { get; set; }

    public Guid? ProductionCompanyId { get; set; }
    public virtual ProductionCompany ProductionCompany { get; set; }

    public virtual ICollection<Genre> Genres { get; set; }
    public virtual ICollection<Screening> Screenings { get; set; }
    public virtual ICollection<ProductionCountry> ProductionCountries { get; set; }
    public virtual ICollection<MovieLanguage> MovieLanguages { get; set; }
    public virtual ICollection<MovieImage> MovieImages { get; set; }
}

File: MovieTheaterUI/src/app/shared/models/movie.ts
import { IGenre } from "./genre";
import { IMovieDirectorModel } from "./movieDirector";
import { IMovieLanguageModel } from "./movieLanguage";
import { IProductionCompany } from "./productionCompany";
import { IProductionCountryModel } from "./productionCountry";
import { IScreeningModel } from "./screening";

export interface IMovieModel{
  movies: IMovie[];
}

export interface IMovie{
  id: string;
  title: string;
  description: string;
  durationMinutes: number;
  dateReleased: Date;
  budget: number;
  imDbRating: number;
  revenue: number;
  poster: string;
  productionCompany: IProductionCompany;
  movieDirector: IMovieDirectorModel;
  screenings: IScreeningModel[];
  countries: IProductionCountryModel[];
  languages: IMovieLanguageModel[];
  genres: IGenre[];
}

export interface ICreateMovie{
  title: string;
  description: string;
  durationMinutes: number;
  dateReleased: Date;
  budget: number;
  imdbRating: number;
  revenue: number;
  poster: string;
}

export interface IEditMovie{
  title: string;
  description: string;
  durationMinutes: number;
  dateReleased: Date;
  budget: number;
  imDbRating: number;
  revenue: number;
  poster: string;
  genres: string[];
  countries: string[];
  languages: string[];
}

export interface IMovieFilter{
  title: string;
  genres: string[];
  runtime: number[];
  companies: string[];
  countries: string[];
  sortingOption: string;
}

export class MovieFilter implements IMovieFilter{
  title: string;
  genres: string[] = [];
  runtime: number[] = [];
  companies: string[] = [];
  countries: string[] = [];
  sortingOption: string;
}
